src/client/opamSolution.ml

```diff
@@ -1129,58 +1129,76 @@ let syspkgs_to_string spkgs =
    true. *)
-let get_depexts ?(force=false) ?(recover=false) t packages =
-  if not force && OpamStateConfig.(!r.no_depexts) then OpamSysPkg.Set.empty else
-  let sys_packages =
-    if recover then
-      OpamSwitchState.depexts_status_of_packages t packages
-    else
-      let base = Lazy.force t.sys_packages in
-      (* workaround: st.sys_packages is not always updated with added
-         packages *)
-      let more_pkgs =
-        OpamPackage.Set.filter (fun nv ->
-            (* dirty heuristic: recompute for all non-canonical packages *)
-            OpamPackage.Map.find_opt nv t.repos_package_index
-            <> OpamSwitchState.opam_opt t nv)
-          packages
-      in
-      if OpamPackage.Set.is_empty more_pkgs then base else
-      OpamPackage.Map.union (fun _ x -> x) base
-        (OpamSwitchState.depexts_status_of_packages t more_pkgs)
-  in
-  let avail, nf =
-    OpamPackage.Set.fold (fun pkg (avail,nf) ->
-        match OpamPackage.Map.find_opt pkg sys_packages with
-        | Some sys ->
-          OpamSysPkg.(Set.union avail sys.s_available),
-          OpamSysPkg.(Set.union nf sys.s_not_found)
-        | None -> avail, nf)
-      packages (OpamSysPkg.Set.empty, OpamSysPkg.Set.empty)
-  in
-  print_depext_msg (avail, nf);
-  avail
+let get_depexts ?(force=false) ?(recover=false) t ~pkg_to_install ~pkg_installed =
+  if not force && OpamStateConfig.(!r.no_depexts) then
+    OpamSysPkg.to_install_empty
+  else
+    let sys_packages =
+      if recover then
+        OpamSwitchState.depexts_status_of_packages t pkg_to_install
+      else
+        let base = Lazy.force t.sys_packages in
+        (* workaround: st.sys_packages is not always updated with added
+           packages *)
+        let more_pkgs =
+          OpamPackage.Set.filter (fun nv ->
+              (* dirty heuristic: recompute for all non-canonical packages *)
+              OpamPackage.Map.find_opt nv t.repos_package_index
+              <> OpamSwitchState.opam_opt t nv)
+            pkg_to_install
+        in
+        if OpamPackage.Set.is_empty more_pkgs then base else
+          OpamPackage.Map.union (fun _ x -> x) base
+            (OpamSwitchState.depexts_status_of_packages t more_pkgs)
+    in
+    let already_installed = OpamPackage.Set.diff pkg_installed pkg_to_install in
+    let status =
+      let open OpamSysPkg.Set.Op in
+      OpamPackage.Set.fold (fun pkg (acc : OpamSysPkg.status) ->
+          match OpamPackage.Map.find_opt pkg sys_packages with
+          | Some sys ->
+            { OpamSysPkg.
+              s_available = acc.s_available ++ sys.s_available;
+              s_not_found = acc.s_not_found ++ sys.s_not_found;
+            }
+          | None -> acc)
+        pkg_to_install OpamSysPkg.status_empty
+    in
+    print_depext_msg status;
+    let ti_required =
+      let bypass = t.switch_config.OpamFile.Switch_config.depext_bypass in
+      OpamPackage.Set.fold OpamSysPkg.Set.Op.(fun nv req ->
```

Comment:
This open can be removed if the same open just above was moved a couple lines above in the function
```suggestion
      OpamPackage.Set.fold (fun nv req ->
```